File: shared/src/commonMain/kotlin/space/denhan/data/Mappers.kt
package space.denhan.data

import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalTime
import space.denhan.db.HandledEventRow
import space.denhan.db.ItemRow
import space.denhan.db.ItemGroupRow
import space.denhan.db.SimProfileRow
import space.denhan.domain.Carrier
import space.denhan.domain.Cycle
import space.denhan.domain.DateSource
import space.denhan.domain.GroupKind
import space.denhan.domain.HandledEvent
import space.denhan.domain.ItemGroup
import space.denhan.domain.ItemState
import space.denhan.domain.Kind
import space.denhan.domain.NagPolicy
import space.denhan.domain.SimProfile
import space.denhan.domain.Stake
import space.denhan.domain.TrackedItem
import space.denhan.domain.TriState

/**
 * Row-to-model conversion.
 *
 * Enums are stored as their names rather than ordinals so that reordering an
 * enum cannot silently reinterpret existing rows, and so the database stays
 * legible in a sqlite3 shell. Unknown values fall back to a safe default rather
 * than throwing: a row written by a newer build must not crash an older one.
 */

private inline fun <reified T : Enum<T>> String.toEnumOr(fallback: T): T =
    enumValues<T>().firstOrNull { it.name == this } ?: fallback

private fun String.toDate(): LocalDate = LocalDate.parse(this)
private fun String?.toDateOrNull(): LocalDate? = this?.let { LocalDate.parse(it) }

private fun String.toLeadDays(): List<Int> =
    if (isBlank()) emptyList() else split(',').mapNotNull { it.trim().toIntOrNull() }

private fun List<Int>.encodeLeadDays(): String = joinToString(",")

fun ItemRow.toDomain() = TrackedItem(
    id = id,
    name = name,
    groupId = groupId,
    kind = kind.toEnumOr(Kind.RECURRING),
    stake = stake.toEnumOr(Stake.INFO),
    categoryId = categoryId,
    expiresOn = expiresOn.toDate(),
    actByOffsetDays = actByOffsetDays.toInt(),
    anchorDate = anchorDate.toDate(),
    cycle = cycle?.let { c -> Cycle.entries.firstOrNull { it.name == c } },
    amountMinor = amountMinor,
    currency = currency,
    actionUrl = actionUrl,
    actionLabel = actionLabel,
    note = note,
    leadDays = leadDays.toLeadDays(),
    remindAt = LocalTime.parse(remindAt),
    nagAfterDue = nagAfterDue.toEnumOr(NagPolicy.NONE),
    verifyEveryDays = verifyEveryDays?.toInt(),
    lastVerifiedAt = lastVerifiedAt.toDateOrNull(),
    // An unrecognised source must read as less trustworthy, not more.
    dateSource = dateSource.toEnumOr(DateSource.USER_ESTIMATED),
    state = state.toEnumOr(ItemState.ACTIVE),
)

fun ItemGroupRow.toDomain() = ItemGroup(
    id = id,
    name = name,
    kind = kind.toEnumOr(GroupKind.OTHER),
)

fun SimProfileRow.toDomain() = SimProfile(
    groupId = groupId,
    carrier = carrier.toEnumOr(Carrier.OTHER),
    msisdn = msisdn,
    isDormant = isDormant != 0L,
    hsdConfirmedDate = hsdConfirmedDate.toDateOrNull(),
    hsdConfirmedAt = hsdConfirmedAt.toDateOrNull(),
    // Anything unrecognised means "we have not established this", never YES.
    identityVerified = identityVerified.toEnumOr(TriState.UNKNOWN),
    identityCheckedAt = identityCheckedAt.toDateOrNull(),
    lastDeviceChangeAt = lastDeviceChangeAt.toDateOrNull(),
    retentionPackage = retentionPackage,
    retentionExpiryDate = retentionExpiryDate.toDateOrNull(),
)

fun HandledEventRow.toDomain() = HandledEvent(
    id = id,
    itemId = itemId,
    handledAtEpochSeconds = handledAt,
    forDueDate = forDueDate.toDate(),
    amountMinor = amountMinor,
    currency = currency,
    fxRateScaled = fxRateScaled,
    fxRateScale = fxRateScale?.toInt(),
    fxRateDate = fxRateDate.toDateOrNull(),
    fxSource = fxSource,
    baseAmountMinor = baseAmountMinor,
    actualChargedMinor = actualChargedMinor,
)

internal fun TrackedItem.encodedLeadDays() = leadDays.encodeLeadDays()

File: lib/data/mappers.dart
import 'package:drift/drift.dart';

import 'package:subdock/domain/local_date.dart';
import 'package:subdock/domain/model.dart';

import 'database.dart';

/// Row-to-model conversion.
///
/// Enums are stored as their wire names rather than ordinals so that reordering
/// an enum cannot silently reinterpret existing rows, and so the database stays
/// legible in a sqlite3 shell. Unknown values fall back to a safe default rather
/// than throwing: a row written by a newer build must not crash an older one.

List<int> decodeLeadDays(String raw) {
  if (raw.trim().isEmpty) return const [];
  return raw
      .split(',')
      .map((part) => int.tryParse(part.trim()))
      .whereType<int>()
      .toList(growable: false);
}

String encodeLeadDays(List<int> leadDays) => leadDays.join(',');

extension CategoryRowMapper on CategoryRowData {
  Category toDomain() => Category(
    id: id,
    label: label,
    iconName: iconName,
    // An unrecognised wording reads as money owed, which is what all but three
    // shipped shelves are. Being told a payment is due when nothing will be
    // taken is a wasted glance; being told nothing when a passport lapses is
    // the failure the app exists to prevent.
    wording: enumFromWire(CategoryWording.values, wording, CategoryWording.due),
    // An unrecognised nag policy stays quiet rather than nagging daily. A shelf
    // written by a newer build must not turn into an alarm on an older one.
    nag: enumFromWire(NagPolicy.values, nag, NagPolicy.none),
    leadDays: decodeLeadDays(leadDays),
    verifyEveryDays: verifyEveryDays,
    countsTowardSpend: countsTowardSpend,
    builtIn: builtIn,
    sortOrder: sortOrder,
  );
}

extension CategoryToRow on Category {
  CategoryRowCompanion toCompanion() => CategoryRowCompanion(
    id: Value(id),
    label: Value(label),
    iconName: Value(iconName),
    wording: Value(wording.wireName),
    nag: Value(nag.wireName),
    leadDays: Value(encodeLeadDays(leadDays)),
    verifyEveryDays: Value(verifyEveryDays),
    countsTowardSpend: Value(countsTowardSpend),
    builtIn: Value(builtIn),
    sortOrder: Value(sortOrder),
  );
}

extension ItemRowMapper on ItemRowData {
  TrackedItem toDomain() => TrackedItem(
    id: id,
    name: name,
    categoryId: category,
    iconName: iconName,
    expiresOn: LocalDate.parse(expiresOn),
    actByOffsetDays: actByOffsetDays,
    anchorDate: LocalDate.parse(anchorDate),
    cycle: CycleWire.fromWire(cycle),
    repeatCount: repeatCount,
    amountMinor: amountMinor,
    currency: currency,
    actionUrl: actionUrl,
    actionLabel: actionLabel,
    note: note,
    leadDays: decodeLeadDays(leadDays),
    remindAt: LocalTime.parse(remindAt),
    nagAfterDue: enumFromWire(NagPolicy.values, nagAfterDue, NagPolicy.none),
    verifyEveryDays: verifyEveryDays,
    lastVerifiedAt: lastVerifiedAt == null
        ? null
        : LocalDate.parse(lastVerifiedAt!),
    // An unrecognised source must read as less trustworthy, not more.
    dateSource: enumFromWire(
      DateSource.values,
      dateSource,
      DateSource.userEstimated,
    ),
    snoozedUntil: snoozedUntil == null ? null : LocalDate.parse(snoozedUntil!),
    state: enumFromWire(ItemState.values, state, ItemState.active),
    // An unrecognised channel must fall back to "we do not know", never to a
    // guess: pointing someone at the wrong billing page wastes the trip.
    purchaseChannel: enumFromWire(
      PurchaseChannel.values,
      purchaseChannel,
      PurchaseChannel.unknown,
    ),
    inTrial: inTrial,
    paymentSourceId: paymentSourceId,
    paused: paused,
    // Undecided is the safe fallback: an unrecognised value shows the
    // suggestion again, which is a mild annoyance. Falling back to `skipped`
    // would silently hide money the user could save.
    yearlyChoice: enumFromWire(
      YearlyChoice.values,
      yearlyChoice,
      YearlyChoice.undecided,
    ),
  );
}

extension PaymentSourceRowMapper on PaymentSourceRowData {
  PaymentSource toDomain() => PaymentSource(
    id: id,
    name: name,
    // An unrecognised glyph draws a card rather than nothing. The name is what
    // identifies the source; the mark is decoration, and a blank square beside
    // a name reads as a bug.
    glyph: enumFromWire(SourceGlyph.values, glyph, SourceGlyph.card),
  );
}

extension HandledEventRowMapper on HandledEventRowData {
  HandledEvent toDomain() => HandledEvent(
    id: id,
    itemId: itemId,
    handledAtEpochSeconds: handledAt,
    forDueDate: LocalDate.parse(forDueDate),
    amountMinor: amountMinor,
    currency: currency,
    fxRateScaled: fxRateScaled,
    fxRateScale: fxRateScale,
    fxRateDate: fxRateDate == null ? null : LocalDate.parse(fxRateDate!),
    fxSource: fxSource,
    baseAmountMinor: baseAmountMinor,
    actualChargedMinor: actualChargedMinor,
  );
}
